drivers/dose/dose.c

```diff
@@ -0,0 +1,572 @@
+            if (esc) {
+                CLRBIT(ctx->flags, DOSE_FLAG_ESC_RECEIVED);
+            }
+            /* Since the dirty flag is set after the RECV state is left,
+             * it indicates that the receive buffer contains unprocessed data
+             * from a previously received frame. Thus, we just ignore new data. */
+            if (!(ctx->flags & DOSE_FLAG_RECV_BUF_DIRTY)
+                && ctx->recv_buf_ptr < DOSE_FRAME_LEN) {
+                ctx->recv_buf[ctx->recv_buf_ptr++] = ctx->uart_octet;
+            }
+        }
+    }
+
+    if (next_state == DOSE_STATE_RECV) {
+        /* Start the octet timeout timer if we are staying in RECV state. */
+        xtimer_set(&ctx->timeout, ctx->timeout_ticks);
+    }
+
+    return next_state;
+}
+
+static uint8_t state_send(dose_t *ctx, uint8_t old_state)
+{
+    if (old_state != DOSE_STATE_SEND) {
+        /* Disable GPIO IRQs during the transmission. */
+        gpio_irq_disable(ctx->sense_pin);
+    }
+
+    /* Don't trace any END octets ... the timeout or the END signal
+     * will bring us back to the BLOCKED state after _send has emitted
+     * its last octet. */
+
+    xtimer_set(&ctx->timeout, ctx->timeout_ticks);
+
+    return DOSE_STATE_SEND;
+}
+
+static uint8_t state(dose_t *ctx, uint8_t src)
+{
+    /* Make sure no other thread or ISR interrupts state transitions */
+    int irq_state = irq_disable();
+
+    uint8_t old_state = ctx->state;
+    uint8_t new_state = DOSE_STATE_UNDEF;
+
+    /* State change based on current state and the input signal.
+     * Since the state only occupies the first 4 bits and the signal the
+     * last 4 bits of a uint8_t, they can be added together and hance
+     * be checked together. */
+    switch (old_state + src) {
+        case DOSE_STATE_UNDEF + DOSE_SIGNAL_INIT:
+            new_state = DOSE_STATE_BLOCKED;
+            break;
+
+        case DOSE_STATE_BLOCKED + DOSE_SIGNAL_GPIO:
+            new_state = DOSE_STATE_RECV;
+            break;
+        case DOSE_STATE_BLOCKED + DOSE_SIGNAL_XTIMER:
+            new_state = DOSE_STATE_IDLE;
+            break;
+
+        case DOSE_STATE_IDLE + DOSE_SIGNAL_GPIO:
+            new_state = DOSE_STATE_RECV;
+            break;
+        case DOSE_STATE_IDLE + DOSE_SIGNAL_SEND:
+            new_state = DOSE_STATE_SEND;
+            break;
+
+        case DOSE_STATE_RECV + DOSE_SIGNAL_UART:
+            new_state = DOSE_STATE_RECV;
+            break;
+        case DOSE_STATE_RECV + DOSE_SIGNAL_XTIMER:
+            new_state = DOSE_STATE_BLOCKED;
+            break;
+
+        case DOSE_STATE_SEND + DOSE_SIGNAL_UART:
+            new_state = DOSE_STATE_SEND;
+            break;
+        case DOSE_STATE_SEND + DOSE_SIGNAL_XTIMER:
+            new_state = DOSE_STATE_BLOCKED;
+            break;
+        case DOSE_STATE_SEND + DOSE_SIGNAL_END:
+            new_state = DOSE_STATE_BLOCKED;
+            break;
+    }
+
+    assert(new_state > DOSE_STATE_UNDEF && new_state < DOSE_STATE_INVALID);
```

Comment:
In rare cases, this assertions actually raises kernel panics! This happens if sense pin interrupts are lost or being overtaken by uart interrupts. I need to rethink the state machine to handle these cases!